### PR TITLE
Fix leading & operators in flow types

### LINKED ISSUE #702: Disable leading & for flow types
```js
type ResponseMessage = MessageCommon & {
  messageType: 'RESPONSE',
  sequenceID: 'number',
};
```

turns into

```js
type ResponseMessage =
  & MessageCommon
  & {
    messageType: "RESPONSE",
    sequenceID: "number"
  };
```

which looks really weird. For `|`, they are usually chained so it feels nice, but you only have maximum one `&` so we should just display it inline like we do with #692

### PR DESCRIPTION
**Pull request by yamafaktory**
This one should fix #702.

### CODE REVIEW

**Comment by yamafaktory** [comment]
I've implemented an adaptive logic to match both @vjeux's expectations and another case that we can discuss:

```js
type ResponseMessage = MessageCommon & {
  messageType: "RESPONSE",
  sequenceID: "number"
};
```

Now turns into (conservative):

```js
type ResponseMessage = MessageCommon & {
  messageType: "RESPONSE",
  sequenceID: "number"
};
```

But when we have a leading type annotation of type `ObjectTypeAnnotation`, I thought keeping the old behavior was better:

```js
type ResponseMessage = {
  messageType: 'RESPONSE',
  sequenceID: 'number',
} & MessageCommon;
```

Turns into:

```js
type ResponseMessage =
  & {
    messageType: "RESPONSE",
    sequenceID: "number"
  }
  & MessageCommon;
```

File: src/printer.js [hunk 1]
```diff
@@ -1539,7 +1539,10 @@ function genericPrintNoParens(path, options, print) {
     case "IntersectionTypeAnnotation":
     case "UnionTypeAnnotation": {
       const types = path.map(print, "types");
-      const op = n.type === "IntersectionTypeAnnotation" ? "&" : "|";
+      const isIntersection = n.type === "IntersectionTypeAnnotation"
+      const op = isIntersection ? "&" : "|";
```
Comment (vjeux) [inline src/printer.js]: you only use `op` where `isIntersection` is true, so you don't need this variable

File: src/printer.js [hunk 2]
```diff
@@ -1548,6 +1551,12 @@ function genericPrintNoParens(path, options, print) {
       // | A
       // | B
       // | C
+
+      // We want & operators to be inlined.
+      if (shouldInline) {
+        return join(concat([" ", op, " "]), types)
```
Comment (vjeux) [inline src/printer.js]: `concat([" ", op, " "])` => `" & "`

File: src/printer.js [hunk 3]
```diff
@@ -1539,7 +1539,10 @@ function genericPrintNoParens(path, options, print) {
     case "IntersectionTypeAnnotation":
     case "UnionTypeAnnotation": {
       const types = path.map(print, "types");
-      const op = n.type === "IntersectionTypeAnnotation" ? "&" : "|";
+      const isIntersection = n.type === "IntersectionTypeAnnotation"
+      const op = isIntersection ? "&" : "|";
+      const shouldInline = isIntersection &&
+        !(n.types.length > 1 && n.types[0].type === "ObjectTypeAnnotation")
```
Comment (vjeux) [inline src/printer.js]: why do you need this condition?

File: tests/comments/__snapshots__/jsfmt.spec.js.snap [hunk 4]
```diff
@@ -384,13 +384,11 @@ const rootEpic = (actions, store) => combineEpics(...epics)(actions, store)
...
 
 // Two extra levels of indentation because of the comment
-export type AsyncExecuteOptions =
-  & child_process$execFileOpts
-  & {
-    // The contents to write to stdin.
-    stdin?: ?string,
-    dontLogInNuclide?: ?boolean
-  };
+export type AsyncExecuteOptions = child_process$execFileOpts & {
```
Comment (vjeux) [inline tests/comments/__snapshots__/jsfmt.spec.js.snap]: All those examples are soooo much better!

File: src/printer.js (hunk 3 above)
Comment (vjeux) [inline src/printer.js]: Okay, nevermind I saw your comment. Works for me.

**Comment by yamafaktory** [comment]
@vjeux I've updated the implementation after your review, thanks!

**Comment by jlongster** [comment]
Looks great! I'll let @vjeux merge it since he did the review :)

**Pull request merged**